File: sdk/python/src/jevql/engine.py
"""Embedded engine: a private ``jevql serve`` process owned by this client.

The binary is discovered in this order: an explicit path, ``JEVQL_ENGINE_PATH``,
the copy bundled inside this package (platform wheels), then ``jevql`` on PATH.
"""

from __future__ import annotations

import atexit
import json
import os
import secrets
import shutil
import subprocess
import sys
import threading
import time
from pathlib import Path
from typing import Any, Dict, List, Optional

from .errors import JevqlError

READY_TIMEOUT = 15.0
STOP_GRACE = 2.0

INSTALL_HINT = (
    "jevql engine binary not found. `pip install jevql` should have included one for this "
    "platform; if you installed from source, set JEVQL_ENGINE_PATH or install the CLI with "
    "`brew install kylemclaren/tap/jevql`."
)


def bundled_engine_path() -> Path:
    name = "jevql.exe" if sys.platform == "win32" else "jevql"
    return Path(__file__).resolve().parent / "_engine" / name


def resolve_engine(engine_path: Optional[str] = None) -> str:
    """Return the path of the engine binary, or raise a ``transport`` error."""
    candidates: List[str] = []
    if engine_path:
        candidates.append(engine_path)
    env = os.environ.get("JEVQL_ENGINE_PATH")
    if env:
        candidates.append(env)
    bundled = bundled_engine_path()
    if bundled.is_file():
        candidates.append(str(bundled))
    on_path = shutil.which("jevql")
    if on_path:
        candidates.append(on_path)
    for c in candidates:
        if os.path.isfile(c) and os.access(c, os.X_OK):
            return c
    if candidates:
        raise JevqlError(f"jevql engine not usable at {candidates[0]!r}: not an executable file", code="transport")
    raise JevqlError(INSTALL_HINT, code="transport")


class Engine:
    """Spawns and owns one ``jevql serve`` process."""

    def __init__(
        self,
        *,
        engine_path: Optional[str] = None,
        database_url: Optional[str] = None,
        api_key: Optional[str] = None,
        api_url: Optional[str] = None,
        model: Optional[str] = None,
        threshold: Optional[float] = None,
        max_rows: Optional[int] = None,
        cache_path: Optional[str] = None,
        no_cache: bool = False,
        ready_timeout: float = READY_TIMEOUT,
    ):
        self.engine_path = engine_path
        self.database_url = database_url
        self.api_key = api_key
        self.api_url = api_url
        self.model = model
        self.threshold = threshold
        self.max_rows = max_rows
        self.cache_path = cache_path
        self.no_cache = no_cache
        self.ready_timeout = ready_timeout
        self.token = secrets.token_urlsafe(24)
        self.url: Optional[str] = None
        self.version: Optional[str] = None
        self.process: Optional[subprocess.Popen] = None
        self._lock = threading.Lock()
        self._registered = False

    def argv(self) -> List[str]:
        binary = resolve_engine(self.engine_path)
        args = [
            binary,
            "serve",
            "--listen",
            "127.0.0.1:0",
            "--token",
            self.token,
            "--ready-json",
            "--parent-pid",
            str(os.getpid()),
        ]
        if self.api_key:
            args += ["--api-key", self.api_key]
        if self.api_url:
            args += ["--api-url", self.api_url]
        if self.model:
            args += ["--model", self.model]
        if self.threshold is not None:
            args += ["--threshold", str(float(self.threshold))]
        if self.max_rows is not None:
            args += ["--max-rows", str(int(self.max_rows))]
        if self.cache_path:
            args += ["--cache", self.cache_path]
        if self.no_cache:
            args.append("--no-cache")
        if self.database_url:
            args.append(self.database_url)
        return args

    @property
    def running(self) -> bool:
        return self.process is not None and self.process.poll() is None

    def start(self) -> str:
        """Start the engine if needed and return its base URL."""
        with self._lock:
            if self.running and self.url:
                return self.url
            argv = self.argv()
            try:
                proc = subprocess.Popen(
                    argv,
                    stdin=subprocess.DEVNULL,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    env=os.environ.copy(),
                    text=True,
                )
            except OSError as e:
                raise JevqlError(f"failed to start jevql engine ({argv[0]}): {e}", code="transport") from None
            self.process = proc
            if not self._registered:
                atexit.register(self.stop)
                self._registered = True
            ready = self._read_ready(proc)
            self.url = str(ready.get("url") or f"http://{ready.get('listen')}")
            self.version = str(ready.get("version") or "")
            return self.url

    def _read_ready(self, proc: subprocess.Popen) -> Dict[str, Any]:
        result: Dict[str, Any] = {}
        err: List[str] = []

        def reader() -> None:
            assert proc.stdout is not None
            line = proc.stdout.readline()
            if line:
                try:
                    payload = json.loads(line)
                except ValueError:
                    err.append(f"unexpected engine output: {line.strip()!r}")
                    return
                if isinstance(payload, dict) and payload.get("ready"):
                    result.update(payload)
                else:
                    err.append(f"unexpected engine output: {line.strip()!r}")

        t = threading.Thread(target=reader, daemon=True)
        t.start()
        deadline = time.monotonic() + self.ready_timeout
        while t.is_alive() and time.monotonic() < deadline:
            if proc.poll() is not None:
                break
            t.join(0.05)
        if result:
            return result
        exited = proc.poll() is not None
        status = proc.returncode
        tail = self._stderr_tail(proc)
        self.stop()
        if exited:
            raise JevqlError(f"jevql engine exited with status {status} before it was ready: {tail}", code="transport")
        reason = err[0] if err else f"no ready line within {self.ready_timeout:.0f}s"
        raise JevqlError(f"jevql engine did not start: {reason}. {tail}".strip(), code="transport")

    @staticmethod
    def _stderr_tail(proc: subprocess.Popen) -> str:
        if proc.stderr is None:
            return ""
        try:
            if proc.poll() is None:
                proc.terminate()
                proc.wait(timeout=STOP_GRACE)
            data = proc.stderr.read() or ""
        except Exception:  # pragma: no cover - best effort
            return ""
        lines = [ln for ln in data.strip().splitlines() if ln.strip()]
        return " | ".join(lines[-5:])

    def stop(self) -> None:
        proc = self.process
        if proc is None:
            return
        self.process = None
        self.url = None
        if proc.poll() is None:
            try:
                proc.terminate()
                proc.wait(timeout=STOP_GRACE)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=STOP_GRACE)
            except OSError:
                pass
        for stream in (proc.stdout, proc.stderr):
            try:
                if stream:
                    stream.close()
            except OSError:  # pragma: no cover
                pass

File: sdk/typescript/src/engine.ts
import { spawn, type ChildProcess } from "node:child_process"
import { randomBytes } from "node:crypto"
import { accessSync, constants, existsSync } from "node:fs"
import { createRequire } from "node:module"
import { delimiter, join } from "node:path"
import { JevqlError } from "./error.js"
import type { EmbeddedOptions } from "./types.js"

/* ─────────────────────────────────────────────────────────
 * ENGINE
 * Finds the jevql binary and runs `jevql serve` as a private,
 * token-protected engine on a loopback port for the lifetime
 * of this process. See sdk/PROTOCOL.md, "Embedded engine".
 * ───────────────────────────────────────────────────────── */

export const PLATFORM_PACKAGES = ["darwin-arm64", "darwin-x64", "linux-x64", "linux-arm64"] as const

/** Name of the npm package that carries the binary for this platform, or null when unsupported. */
export function platformPackage(platform: string = process.platform, arch: string = process.arch): string | null {
  const key = `${platform}-${arch}`
  return (PLATFORM_PACKAGES as readonly string[]).includes(key) ? `@jevql/engine-${key}` : null
}

function executable(p: string): boolean {
  try {
    accessSync(p, constants.X_OK)
    return true
  } catch {
    return false
  }
}

export interface ResolveOptions {
  enginePath?: string
  env?: NodeJS.ProcessEnv
  /** Resolver for the bundled package (injectable for tests). */
  resolve?: (specifier: string) => string
  platform?: NodeJS.Platform
  arch?: string
}

/** Locate the jevql binary: enginePath, JEVQL_ENGINE_PATH, bundled platform package, then PATH. */
export function resolveEngine(opts: ResolveOptions = {}): string {
  const env = opts.env ?? process.env
  if (opts.enginePath) {
    if (!existsSync(opts.enginePath)) throw new JevqlError(`jevql: enginePath ${opts.enginePath} does not exist`, "transport")
    return opts.enginePath
  }
  if (env.JEVQL_ENGINE_PATH) {
    if (!existsSync(env.JEVQL_ENGINE_PATH)) throw new JevqlError(`jevql: JEVQL_ENGINE_PATH ${env.JEVQL_ENGINE_PATH} does not exist`, "transport")
    return env.JEVQL_ENGINE_PATH
  }
  const pkg = platformPackage(opts.platform, opts.arch)
  if (pkg) {
    const resolve = opts.resolve ?? createRequire(import.meta.url).resolve
    try {
      const p = resolve(`${pkg}/bin/jevql`)
      if (executable(p)) return p
    } catch {
      /* not installed */
    }
  }
  const exe = (opts.platform ?? process.platform) === "win32" ? "jevql.exe" : "jevql"
  for (const dir of (env.PATH ?? "").split(delimiter)) {
    if (!dir) continue
    const p = join(dir, exe)
    if (executable(p)) return p
  }
  const hint = pkg ? `npm i ${pkg}` : "a platform we do not ship binaries for"
  throw new JevqlError(
    `jevql: engine binary not found. Install it with \`brew install kylemclaren/tap/jevql\`, or \`${hint}\`, or set JEVQL_ENGINE_PATH.`,
    "transport",
  )
}

/** Command-line flags for the engine from explicit options. Exported for tests. */
export function engineArgs(o: EmbeddedOptions, token: string, parentPid = process.pid): string[] {
  const a = ["serve", "--listen", "127.0.0.1:0", "--token", token, "--ready-json", "--parent-pid", String(parentPid)]
  if (o.apiKey) a.push("--api-key", o.apiKey)
  if (o.apiUrl) a.push("--api-url", o.apiUrl)
  if (o.model) a.push("--model", o.model)
  if (o.threshold !== undefined) a.push("--threshold", String(o.threshold))
  if (o.maxRows !== undefined) a.push("--max-rows", String(o.maxRows))
  if (o.cachePath) a.push("--cache", o.cachePath)
  if (o.noCache) a.push("--no-cache")
  if (o.databaseUrl) a.push(o.databaseUrl)
  return a
}

export interface Ready {
  url: string
  version: string
  pid: number
}

const running = new Set<Engine>()
let exitHookInstalled = false
function installExitHook() {
  if (exitHookInstalled) return
  exitHookInstalled = true
  process.on("exit", () => {
    for (const e of running) e.killNow()
  })
}

/** One engine process. */
export class Engine {
  readonly token: string
  private child: ChildProcess | null = null
  private starting: Promise<Ready> | null = null
  private ready: Ready | null = null

  constructor(private readonly opts: EmbeddedOptions) {
    this.token = randomBytes(24).toString("hex")
  }

  /** Bound URL once started. */
  get info(): Ready | null {
    return this.ready
  }

  /** Start (once) and resolve the ready line. Concurrent callers share the same spawn. */
  start(): Promise<Ready> {
    if (this.ready) return Promise.resolve(this.ready)
    if (!this.starting) {
      this.starting = this.spawn().then(
        (r) => {
          this.ready = r
          return r
        },
        (e) => {
          this.starting = null
          throw e
        },
      )
    }
    return this.starting
  }

  private spawn(): Promise<Ready> {
    const binary = resolveEngine({ enginePath: this.opts.enginePath, env: { ...process.env, ...(this.opts.env ?? {}) } })
    const args = engineArgs(this.opts, this.token)
    const env = { ...process.env, ...(this.opts.env ?? {}) }
    const timeoutMs = (this.opts.startTimeout ?? 15) * 1000
    return new Promise<Ready>((resolve, reject) => {
      let child: ChildProcess
      try {
        child = spawn(binary, args, { env, stdio: ["ignore", "pipe", "pipe"] })
      } catch (e) {
        reject(new JevqlError(`jevql: cannot start engine ${binary}: ${(e as Error).message}`, "transport"))
        return
      }
      this.child = child
      running.add(this)
      installExitHook()
      let out = ""
      let err = ""
      let done = false
      const fail = (msg: string) => {
        if (done) return
        done = true
        clearTimeout(timer)
        this.killNow()
        const tail = err.trim().split("\n").slice(-6).join("\n")
        reject(new JevqlError(`jevql: ${msg}${tail ? `\n${tail}` : ""}`, "transport"))
      }
      const timer = setTimeout(() => fail(`engine did not report ready within ${timeoutMs / 1000}s`), timeoutMs)
      child.stdout!.setEncoding("utf8")
      child.stderr!.setEncoding("utf8")
      child.stderr!.on("data", (d: string) => {
        err += d
        if (err.length > 64 * 1024) err = err.slice(-32 * 1024)
      })
      child.stdout!.on("data", (d: string) => {
        if (done) return
        out += d
        const nl = out.indexOf("\n")
        if (nl < 0) return
        const line = out.slice(0, nl).trim()
        try {
          const j = JSON.parse(line) as { ready?: boolean; url?: string; listen?: string; version?: string; pid?: number }
          if (!j.ready || !(j.url || j.listen)) throw new Error("bad ready line")
          done = true
          clearTimeout(timer)
          resolve({ url: j.url ?? `http://${j.listen}`, version: j.version ?? "", pid: j.pid ?? child.pid ?? 0 })
        } catch {
          fail(`engine printed something other than a ready line: ${line.slice(0, 200)}`)
        }
      })
      child.on("error", (e) => fail(`cannot start engine ${binary}: ${e.message}`))
      child.on("exit", (code, signal) => {
        running.delete(this)
        this.child = null
        if (!done) fail(`engine exited before ready (${signal ?? `code ${code}`})`)
      })
    })
  }

  /** SIGKILL immediately (used at process exit). */
  killNow(): void {
    const c = this.child
    if (c && c.exitCode === null && !c.killed) {
      try {
        c.kill("SIGKILL")
      } catch {
        /* already gone */
      }
    }
    running.delete(this)
  }

  /** Graceful stop: SIGTERM, wait up to 2s, then SIGKILL. */
  async stop(): Promise<void> {
    const c = this.child
    this.ready = null
    this.starting = null
    running.delete(this)
    if (!c || c.exitCode !== null) {
      this.child = null
      return
    }
    await new Promise<void>((resolve) => {
      const t = setTimeout(() => {
        try {
          c.kill("SIGKILL")
        } catch {
          /* gone */
        }
      }, 2000)
      c.once("exit", () => {
        clearTimeout(t)
        resolve()
      })
      try {
        c.kill("SIGTERM")
      } catch {
        clearTimeout(t)
        resolve()
      }
    })
    this.child = null
  }
}
